### ISSUE TITLE
proposal: `file_import_inside_lib_references_file_outside`

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#4103]
## Description

This rule was originally proposed (and implemented?) as an analyzer Hint, but the implementation is either gone or never existed, and this is more a style recommendation so should be a Lint rule.

## Details

The YAML definition of the Hint was the following:

```yaml
  FILE_IMPORT_INSIDE_LIB_REFERENCES_FILE_OUTSIDE:
    problemMessage: "A file in the 'lib' directory shouldn't import a file outside the 'lib' directory."
    correctionMessage: "Try removing the import, or moving the imported file inside the 'lib' directory."
    comment: |-
      It is a bad practice for a source file in a package "lib" directory
      hierarchy to traverse outside that directory hierarchy. For example, a
      source file in the "lib" directory should not contain a directive such as
      `import '../web/some.dart'` which references a file outside the lib
      directory.
```

## Kind

Style

## Bad Examples

Relative imports from `lib/` that go outside `lib/`.

```dart
// in lib/foo.dart:

import '../bin/bar.dart';
```

## Good Examples 

Relative imports from `lib/` that don't go outside `lib/`.

```dart
// in lib/foo.dart:

import 'bar.dart';
```

Relative imports from outside `lib/`.

```dart
// in test/foo.dart:

import '../tool/bar.dart';
```

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [ ] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [ ] If there's any prior art (e.g., in other linters), please add references here.
- [ ] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [ ] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by srawlins** [comment]
FWIW, I am proposing this primarily as due diligence in deleting the (not implemented) analyzer Hint. But also I would support this rule. I don't see it happen much, but it's a legitimate request, helpful for new users (as our `lib/` setup is bespoke).
